### PR TITLE
feat(project): observe schema-version classification on IDB project load (#553 Slice B, 1/N)

### PR DESCRIPTION
**Pull request by qnbs**
## Summary

First ingress path of Slice B, wired to the classifiers built in Slice A (#618), per the admitted contract's §2.8 universal ingress admission requirement.

- **`classifyProjectVersionFromObject`** (new, `features/project/projectSchemaVersion.ts`) — a second classifier variant for ingress paths with no raw JSON text to begin with. Verified against live source: IndexedDB stores/retrieves structured-clone JS objects directly (`IDBObjectStore.put`/`.get`), never a JSON string — `classifyRawProjectVersion`'s raw-text scanning (duplicate-key detection, escape validation) doesn't apply here, and a JS object literal cannot carry duplicate keys in the first place.
- **`projectSchemaVersionShadow.ts`** (new) — the observation wrapper, matching this codebase's established Core shadow-validation pattern (`coreValidationShadow.ts`): classify and log via `services/logger.ts`, never alter load behavior, never throw. Covers both the new object-based classifier and Slice A's text-based one (for the remaining ingress paths in this slice).
- Wired into `IdbProjectStore#loadState` (via `validateAndFixState`), observing the as-persisted project data before any default-backfilling runs.

**Deliberately observation-only, not enforcing:** actual admission gating (refusing `FUTURE`/`MALFORMED`, migrating `LEGACY_UNVERSIONED`) is deferred to the slice that implements §2.4's fail-closed mechanism and the `LEGACY_TO_V1` migration. Every real, existing project has no `schemaVersion` field today — enforcing that migration now, before it exists, would immediately reclassify every user's every project on next load. This PR only proves the classifier against real, live ingress data first.

**Remaining ingress paths** (filesystem load, IDB snapshot restore, backup import, recovery restore) are separate, subsequent Slice B commits/PRs — one causal path at a time.

## Test plan
- [x] `pnpm exec vitest run tests/unit/features/project/projectSchemaVersion.test.ts` — 40/40 passing
- [x] `pnpm exec vitest run tests/unit/features/project/projectSchemaVersionShadow.test.ts` — 6/6 passing
- [x] `pnpm exec vitest run tests/unit/services/storage/idbProjectStoreLoadStateObservation.test.ts` — 2/2 passing (verifies observation fires before default-backfilling, and does not fire when no project is persisted)
- [x] `pnpm run ci:prepush` — local admission gate PASS
- [ ] CI: exact-head required + advisory checks green
- [ ] CI: CodeQL SUCCESS

## Summary by Sourcery

Observe project schema compatibility during IndexedDB loads without changing existing project loading behavior.

New Features:
- Add object-based project schema-version classification for structured-clone persistence data.
- Add observation-only schema-version logging for project ingress paths, including IndexedDB loads.

Bug Fixes:
- Ensure malformed, falsy, and unusual persisted project records are observed without interrupting loading, while absent records remain ignored.

Enhancements:
- Classify persisted IndexedDB project data before default backfilling and surface incompatible versions at warning level without changing admission or load behavior.

Documentation:
- Update documented test counts to reflect the added coverage.

Tests:
- Add coverage for object-based classification, observation behavior, error containment, and IndexedDB load ordering and edge cases.

---
## Summary by cubic
Adds observation-only schema-version classification to IndexedDB project loads, so schema compatibility issues are logged before any default-backfilling runs. Previously no classification happened on load; load behavior is unchanged and observation never throws.

- New `classifyProjectVersionFromObject` classifies structured-clone objects (IndexedDB has no raw JSON text), rejecting non-record clones like `Date` and typed arrays.
- Observation is gated on the record being present (`project !== undefined`), not truthy, so corrupted falsy records are still observed; `MALFORMED`, `FUTURE`, and `UNSUPPORTED_OLDER` log at warn, the rest at debug.
- A record carrying its own top-level `schemaVersion` is classified directly, before the envelope-unwrap heuristic, so a flat `FUTURE` record with an unrelated `data` field isn't misread as `LEGACY_UNVERSIONED`.
- Enforcing admission (refusing `FUTURE`/`MALFORMED`, migrating `LEGACY_UNVERSIONED`) is deferred; every existing project lacks `schemaVersion` today.
- Remaining ingress paths (filesystem load, backup import, snapshot restore) are separate follow-up PRs.

<sup>Written for commit 5871fdd3905c11e68f9a272faf6ddd11d12e394c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/qnbs/WorldScript-Studio/pull/619?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Improvements**
  * Improved project schema-version detection during project loading, including malformed, unsupported, future, legacy, and missing-version data.
  * Added diagnostic logging for compatibility issues without interrupting normal loading or default handling.
  * Extended detection to cover persisted project records in varied or malformed states.

* **Documentation**
  * Updated README test metrics to reflect 7,494+ tests across 601 files.

___

## **CodeAnt-AI Description**
Observe project schema versions during IndexedDB loads without changing load behavior

### What Changed
- IndexedDB project loads now classify the persisted project data before default fields are added.
- Invalid, future, or unsupported schema versions are logged as warnings; normal and legacy projects are logged for observation.
- Corrupt, null, falsy, and unusual persisted records are still observed, while missing project records are ignored.
- Schema classification failures are contained and cannot interrupt project loading.
- Added coverage for object-based schema classification and IndexedDB observation behavior.

### Impact
`✅ Earlier visibility into incompatible project data`
`✅ Project loads continue when observation encounters malformed data`
`✅ Clearer warnings for unsupported schema versions`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: features/project/projectSchemaVersion.ts [hunk 1]
```diff
@@ -186,4 +186,29 @@ export function classifyRawProjectVersion(rawText: string): ProjectVersionClassi
...
+ * contract section 2.8's universal ingress admission names IndexedDB explicitly, and IndexedDB
+ * stores/retrieves structured-clone JS objects directly (`IDBObjectStore.put`/`.get`), with no
+ * JSON serialization step anywhere in that path (verified against
+ * `services/storage/idbProjectStore.ts`). Unlike `classifyRawProjectVersion`, there is no
+ * duplicate-key or JSON-escape concern here: a JS object literal cannot carry two properties with
+ * the same key — structured clone, like `JSON.parse`, has already resolved that before this
+ * function ever sees the value — so this only needs the same value-grammar and
+ * version-comparison rules, not a raw-text scan.
+ */
+export function classifyProjectVersionFromObject(value: unknown): ProjectVersionClassification {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/projectSchemaVersion.ts:202]: Fixed in 32dec21e: added QNBS-v3 comments at all 4 cited sites.

File: services/storage/idbProjectStore.ts [hunk 2]
```diff
@@ -224,6 +225,8 @@ export class IdbProjectStore extends IdbAssetStore {
     if (validProject) {
       const rawData = validProject.present ? validProject.present.data : validProject.data;
       if (rawData) {
+        // QNBS-v3: observes the as-persisted shape before any default-backfilling below, per contract section 2.8's universal ingress admission (observation-only, Slice B).
+        observeProjectVersionClassificationFromObject(rawData, 'idb-project-load');
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/storage/idbProjectStore.ts]: Fixed in bf4568fc: the observation call now falls back to the envelope itself (rawData ?? validProject) so a present but flat/malformed record is never silently skipped. Added a regression test.

File: features/project/projectSchemaVersionShadow.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ * behavior, never throw. Actual admission gating (refusing `FUTURE`/`MALFORMED`, migrating
+ * `LEGACY_UNVERSIONED`) is deferred to the #553 slice that implements section 2.4's fail-closed
+ * mechanism and the `LEGACY_TO_V1` migration — this only proves the classifiers against real,
+ * live ingress data first.
+ */
+function logClassification(
+  classification: ProjectVersionClassification,
+  ingressPath: string,
+): void {
+  const level = classification === 'MALFORMED' || classification === 'FUTURE' ? 'warn' : 'debug';
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/projectSchemaVersionShadow.ts]: Fixed in bf4568fc: UNSUPPORTED_OLDER now logs at warn alongside MALFORMED/FUTURE, since contract section 2.5 treats it identically (no write authority). Added a regression test.

File: features/project/projectSchemaVersion.ts [hunk 4]
```diff
@@ -186,4 +186,29 @@ export function classifyRawProjectVersion(rawText: string): ProjectVersionClassi
...
+  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
+    return 'MALFORMED';
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/projectSchemaVersion.ts:204]: Fixed in bf4568fc: now requires a plain object (Object.prototype or null prototype) before proceeding, rejecting Date/Map/typed-array structured-clone values. Added regression tests for both the rejection and the null-prototype acceptance case.

File: features/project/projectSchemaVersionShadow.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import { logger } from '../../services/logger';
+import {
+  classifyProjectVersionFromObject,
+  classifyRawProjectVersion,
+  type ProjectVersionClassification,
+} from './projectSchemaVersion';
+
+/**
+ * Universal ingress admission (contract section 2.8), staged as observation-only — the first
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/projectSchemaVersionShadow.ts:10]: Fixed in 32dec21e (and reinforced in bf4568fc): added a module-level QNBS-v3 comment.

File: tests/unit/features/project/projectSchemaVersionShadow.test.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        ingressPath: 'idb-load',
+        error: 'boom',
+      },
+    );
+  });
+});
+
+describe('observeProjectVersionClassificationFromText', () => {
+  it('logs CURRENT at debug level', () => {
+    observeProjectVersionClassificationFromText('{"schemaVersion": 1}', 'filesystem-load');
```
[the remark this answers is not in the document]
Comment (qnbs) [reply tests/unit/features/project/projectSchemaVersionShadow.test.ts]: Fixed in bf4568fc: the CURRENT-version test literal now derives from CURRENT_PROJECT_SCHEMA_VERSION instead of a hardcoded 1.

File: services/storage/idbProjectStore.ts [hunk 7]
```diff
@@ -223,6 +224,8 @@ export class IdbProjectStore extends IdbAssetStore {
     // Ensure Project Structure consistency
     if (validProject) {
       const rawData = validProject.present ? validProject.present.data : validProject.data;
+      // QNBS-v3: falls back to the envelope itself so a flat/malformed present record (neither .present.data nor .data resolving) is still observed, not silently skipped, per contract section 2.8's universal ingress admission (observation-only, Slice B).
+      observeProjectVersionClassificationFromObject(rawData ?? validProject, 'idb-project-load');
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/storage/idbProjectStore.ts]: Fixed in `bf4568fc`'s follow-up commit. `observeProjectVersionClassificationFromObject(rawData === undefined ? validProject : rawData, ...)` now replaces the `rawData ?? validProject` nullish-coalescing fallback: it only falls back to the outer envelope when neither `.present.data` nor `.data` resolved *at all* (i.e. `undefined`), and observes an explicit `null`/`false` persisted payload as itself (which `classifyProjectVersionFromObject` correctly reports as `MALFORMED`), not as the envelope. Added a regression test (`tests/unit/services/storage/idbProjectStoreLoadStateObservation.test.ts`, 'observes an explicit null payload as itself, not the outer envelope') covering the `{ data: null }` case.

File: features/project/projectSchemaVersion.ts [hunk 8]
```diff
@@ -186,4 +186,35 @@ export function classifyRawProjectVersion(rawText: string): ProjectVersionClassi
...
+  }
+  // QNBS-v3: IDB can store non-record structured-clone values (Date, Map, typed arrays); only a plain object literal (or Object.create(null)) is a valid project-envelope shape.
+  const prototype = Object.getPrototypeOf(value);
+  if (prototype !== null && prototype !== Object.prototype) {
+    return 'MALFORMED';
+  }
+  const record = value as Record<string, unknown>;
+  if (!Object.hasOwn(record, 'schemaVersion')) return 'LEGACY_UNVERSIONED';
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/projectSchemaVersion.ts:212]: Verified: the compat claim is correct, but this is a pre-existing, codebase-wide characteristic, not a regression introduced by this PR. `vite.config.ts` has set `build.target: 'es2022'` with `polyfill: false` since `21a17d9d` (2026-04-18). `services/appBootstrap.ts`'s `normalizeEntityCollection` already calls `Object.hasOwn` unconditionally on the app-bootstrap hot path (`b0286f0a`, 2026-08-31) — every load with any character/world entity already hard-requires `Object.hasOwn` support, independent of this PR. `features/project/coreBoundaryAdapter.ts`, `features/project/adapters.ts`, `features/project/thunks/projectManagementThunks.ts`, and `components/manuscript/ManuscriptEditor.tsx` also already use it in production. Patching only `projectSchemaVersion.ts` would not close the real gap (the other call sites would still throw) and would leave the README's stated Chrome 90+/Firefox 88+/Safari 14+ floor still wrong either way. Opened qnbs/WorldScript-Studio#620 to track the real fix (raise the documented floor to match ES2022, or replace every `Object.hasOwn` call site with `Object.prototype.hasOwnProperty.call` repo-wide) as its own scoped item rather than a piecemeal fix here.

File: services/storage/idbProjectStore.ts [hunk 9]
```diff
@@ -223,6 +224,11 @@ export class IdbProjectStore extends IdbAssetStore {
     // Ensure Project Structure consistency
     if (validProject) {
       const rawData = validProject.present ? validProject.present.data : validProject.data;
+      // QNBS-v3: undefined (not merely falsy) means neither .present.data nor .data resolved at all, so only then fall back to the envelope - a persisted null/false payload must be observed as itself, per contract section 2.8's universal ingress admission (observation-only, Slice B).
+      observeProjectVersionClassificationFromObject(
+        rawData === undefined ? validProject : rawData,
+        'idb-project-load',
+      );
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/storage/idbProjectStore.ts]: Confirmed and fixed in `d243b28b`. `const validProject = project ? (project as PersistedProjectState) : undefined;` used truthiness, so a present-but-falsy top-level project record (`null`/`false`/`0`/`''`) collapsed to `undefined` and the whole `if (validProject)` block — including the observer call — was skipped, even though this is distinct from an absent IDB record (which is genuinely `undefined`). Moved the observation ahead of the `!project && !settings` early return, gated on `project !== undefined`: for a non-object/null `project`, it now observes `project` itself directly (which `classifyProjectVersionFromObject` correctly reports as `MALFORMED`); for an object-shaped record it keeps the existing `.present.data`/`.data`/envelope-fallback selection. The real load/return behavior (the early-return itself, and the backfill block below it) is untouched — only the observation's reachability changed. Added a regression test ('observes a top-level falsy-but-present project record as itself') covering `project === null` with no settings, asserting both the observation call and that the actual load result is still `undefined` as before.

File: services/storage/idbProjectStore.ts [hunk 10]
```diff
@@ -215,6 +216,23 @@ interface PersistedState {
...
+    // QNBS-v3: project !== undefined (not truthiness) distinguishes an absent IDB record from a present-but-falsy one, so a corrupted top-level project record is observed instead of silently treated as absent.
+    if (project !== undefined) {
+      const projectRecord =
+        typeof project === 'object' && project !== null
+          ? (project as PersistedProjectState)
+          : undefined;
+      const observedRawData = projectRecord
+        ? projectRecord.present
+          ? projectRecord.present.data
+          : projectRecord.data
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/storage/idbProjectStore.ts]: Confirmed and fixed in `5871fdd3`. `selectIdbProjectObservationTarget` unwrapped `.present.data`/`.data` unconditionally whenever present, so a flat/malformed record carrying its OWN valid top-level `schemaVersion` plus an unrelated truthy `data`/`present` field (e.g. `{ schemaVersion: FUTURE, data: {} }`) got classified via that nested field (an empty object, misread as `LEGACY_UNVERSIONED`) instead of its own header (which is genuinely `FUTURE`). Fixed by checking `Object.hasOwn(projectRecord, 'schemaVersion')` first: a record that already declares its own version is self-describing and is classified directly, before any Redux-envelope-unwrap guess is attempted. Added a regression test ('classifies a flat record with its own schemaVersion directly, not via a coincidental data field') covering exactly the `{ schemaVersion: 999, data: {} }` case from the finding.
